File: src/common/web/api/user/UserCommands.ts
import { Type } from "class-transformer";

import { Command } from "../../core/messaging/Command";
import { CommandReply } from "../../core/messaging/CommandReply";
import { CommandComposer } from "../../core/messaging/composers/CommandComposer";
import { CommandReplyComposer } from "../../core/messaging/composers/CommandReplyComposer";
import { MessageBuilder } from "../../core/messaging/composers/MessageBuilder";
import { Message } from "../../core/messaging/Message";
import { AuthorizationState } from "../../data/entities/authorization/AuthorizationState";
import { UserSettings } from "../../data/entities/user/UserSettings";
import { type UserToken } from "../../data/entities/user/UserToken";

/**
 * Command to authenticate a user. Note that the actual login/authentication is performed by the underlying host system. Requires a ``AuthenticateUserReply`` reply.
 */
@Message.define("command/user/authenticate")
export class AuthenticateUserCommand extends Command {
    public readonly user_token: UserToken = { user_id: "", user_name: "" } as UserToken;

    /**
     * Helper function to easily build this message.
     */
    public static build(messageBuilder: MessageBuilder, userToken: UserToken, chain: Message | null = null): CommandComposer<AuthenticateUserCommand> {
        return messageBuilder.buildCommand(AuthenticateUserCommand, { user_token: userToken }, chain);
    }
}

/**
 * Reply to ``AuthenticateUserCommand``.
 *
 * @param authorization_state - The authorization state of the user in his host system.
 */
@Message.define("command/user/authenticate/reply")
export class AuthenticateUserReply extends CommandReply {
    public readonly authorization_state: AuthorizationState = AuthorizationState.NotAuthorized;

    public readonly fingerprint: string = "";

    /**
     * Helper function to easily build this message.
     */
    public static build(
        messageBuilder: MessageBuilder,
        cmd: AuthenticateUserCommand,
        authState: AuthorizationState,
        fingerprint: string = "",
        success: boolean = true,
        message: string = "",
    ): CommandReplyComposer<AuthenticateUserReply> {
        return messageBuilder.buildCommandReply(AuthenticateUserReply, cmd, success, message, { authorization_state: authState, fingerprint: fingerprint });
    }
}

/**
 * Command to get the settings of the current user. Requires a ``GetUserSettingsReply`` reply.
 */
@Message.define("command/user/settings/get")
export class GetUserSettingsCommand extends Command {
    /**
     * Helper function to easily build this message.
     */
    public static build(messageBuilder: MessageBuilder, chain: Message | null = null): CommandComposer<GetUserSettingsCommand> {
        return messageBuilder.buildCommand(GetUserSettingsCommand, {}, chain);
    }
}

/**
 * Reply to ``GetUserSettingsCommand``.
 *
 * @param settings - The user settings.
 */
@Message.define("command/user/settings/get/reply")
export class GetUserSettingsReply extends CommandReply {
    // @ts-ignore
    @Type(() => UserSettings)
    public readonly settings: UserSettings = new UserSettings();

    /**
     * Helper function to easily build this message.
     */
    public static build(
        messageBuilder: MessageBuilder,
        cmd: GetUserSettingsCommand,
        settings: UserSettings,
        success: boolean = true,
        message: string = "",
    ): CommandReplyComposer<GetUserSettingsReply> {
        return messageBuilder.buildCommandReply(GetUserSettingsReply, cmd, success, message, { settings: settings });
    }
}

/**
 * Command to set the settings of the current user. Requires a ``SetUserSettingsReply`` reply.
 *
 * @param settings - The new user settings.
 */
@Message.define("command/user/settings/set")
export class SetUserSettingsCommand extends Command {
    // @ts-ignore
    @Type(() => UserSettings)
    public readonly settings: UserSettings = new UserSettings();

    /**
     * Helper function to easily build this message.
     */
    public static build(messageBuilder: MessageBuilder, settings: UserSettings, chain: Message | null = null): CommandComposer<SetUserSettingsCommand> {
        return messageBuilder.buildCommand(SetUserSettingsCommand, { settings: settings }, chain);
    }
}

/**
 * Reply to ``SetUserSettingsCommand``.
 *
 * @param settings - The new user settings (note that these might have been adjusted by the server).
 */
@Message.define("command/user/settings/set/reply")
export class SetUserSettingsReply extends CommandReply {
    // @ts-ignore
    @Type(() => UserSettings)
    public readonly settings: UserSettings = new UserSettings();

    /**
     * Helper function to easily build this message.
     */
    public static build(
        messageBuilder: MessageBuilder,
        cmd: SetUserSettingsCommand,
        settings: UserSettings,
        success: boolean = true,
        message: string = "",
    ): CommandReplyComposer<SetUserSettingsReply> {
        return messageBuilder.buildCommandReply(SetUserSettingsReply, cmd, success, message, { settings: settings });
    }
}

File: src/common/py/api/user/user_commands.py
import dataclasses

from ...core.messaging import (
    Command,
    CommandReply,
    Message,
)
from ...core.messaging.composers import (
    MessageBuilder,
    CommandComposer,
    CommandReplyComposer,
)
from ...data.entities.authorization import AuthorizationState
from ...data.entities.user import User, UserToken


@Message.define("command/user/authenticate")
class AuthenticateUserCommand(Command):
    """
    Command to authenticate a user. Note that the actual login/authentication is performed by the underlying host system.

    Args:
        user_token: The user token.

    Notes:
        Requires a ``AuthenticateUserReply`` reply.
    """

    user_token: UserToken = dataclasses.field(default_factory=UserToken)

    @staticmethod
    def build(
        message_builder: MessageBuilder,
        *,
        user_token: UserToken,
        chain: Message | None = None,
    ) -> CommandComposer:
        """
        Helper function to easily build this message.
        """
        return message_builder.build_command(
            AuthenticateUserCommand, chain, user_token=user_token
        )


@Message.define("command/user/authenticate/reply")
class AuthenticateUserReply(CommandReply):
    """
    Reply to ``AuthenticateUserCommand``.

    Args:
        authorization_state: The authorization state of the user in his host system.
        fingerprint: The user's fingerprint.
    """

    authorization_state: AuthorizationState

    fingerprint: str

    @staticmethod
    def build(
        message_builder: MessageBuilder,
        cmd: AuthenticateUserCommand,
        *,
        authorization_state: AuthorizationState,
        fingerprint: str = "",
        success: bool = True,
        message: str = "",
    ) -> CommandReplyComposer:
        """
        Helper function to easily build this message.
        """
        return message_builder.build_command_reply(
            AuthenticateUserReply,
            cmd,
            success,
            message,
            authorization_state=authorization_state,
            fingerprint=fingerprint,
        )


@Message.define("command/user/settings/get")
class GetUserSettingsCommand(Command):
    """
    Command to get the settings of the current user.

    Notes:
        Requires a ``GetUserSettingsReply`` reply.
    """

    @staticmethod
    def build(
        message_builder: MessageBuilder, *, chain: Message | None = None
    ) -> CommandComposer:
        """
        Helper function to easily build this message.
        """
        return message_builder.build_command(GetUserSettingsCommand, chain)


@Message.define("command/user/settings/get/reply")
class GetUserSettingsReply(CommandReply):
    """
    Reply to ``GetUserSettingsCommand``.

    Args:
        settings: The user settings.
    """

    settings: User.Settings = dataclasses.field(default_factory=User.Settings)

    @staticmethod
    def build(
        message_builder: MessageBuilder,
        cmd: GetUserSettingsCommand,
        *,
        settings: User.Settings,
        success: bool = True,
        message: str = "",
    ) -> CommandReplyComposer:
        """
        Helper function to easily build this message.
        """
        return message_builder.build_command_reply(
            GetUserSettingsReply,
            cmd,
            success,
            message,
            settings=settings,
        )


@Message.define("command/user/settings/set")
class SetUserSettingsCommand(Command):
    """
    Command to set the settings of the current user.

    Args:
        settings: The new user settings.

    Notes:
        Requires a ``SetUserSettingsReply`` reply.
    """

    settings: User.Settings = dataclasses.field(default_factory=User.Settings)

    @staticmethod
    def build(
        message_builder: MessageBuilder,
        *,
        settings: User.Settings,
        chain: Message | None = None,
    ) -> CommandComposer:
        """
        Helper function to easily build this message.
        """
        return message_builder.build_command(
            SetUserSettingsCommand, chain, settings=settings
        )


@Message.define("command/user/settings/set/reply")
class SetUserSettingsReply(CommandReply):
    """
    Reply to ``SetUserSettingsCommand``.

    Args:
        settings: The new user settings (note that these might have been adjusted by the server).
    """

    settings: User.Settings = dataclasses.field(default_factory=User.Settings)

    @staticmethod
    def build(
        message_builder: MessageBuilder,
        cmd: SetUserSettingsCommand,
        *,
        settings: User.Settings,
        success: bool = True,
        message: str = "",
    ) -> CommandReplyComposer:
        """
        Helper function to easily build this message.
        """
        return message_builder.build_command_reply(
            SetUserSettingsReply,
            cmd,
            success,
            message,
            settings=settings,
        )
